src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -377,49 +378,58 @@ public static byte[] UrlEncodeToBytes(byte[] value, int offset, int count)
 
         private static string UrlDecodeInternal(string value, Encoding encoding)
         {
-            if (value == null)
+            if (string.IsNullOrEmpty(value))
             {
-                return null;
+                return value;
             }
 
             int count = value.Length;
             UrlDecoder helper = new UrlDecoder(count, encoding);
 
-            // go through the string's chars collapsing %XX and
-            // appending each char as char, with exception of %XX constructs
-            // that are appended as bytes
-
-            for (int pos = 0; pos < count; pos++)
+            try
             {
-                char ch = value[pos];
+                // go through the string's chars collapsing %XX and
+                // appending each char as char, with exception of %XX constructs
+                // that are appended as bytes
 
-                if (ch == '+')
-                {
-                    ch = ' ';
-                }
-                else if (ch == '%' && pos < count - 2)
+                for (int pos = 0; pos < count; pos++)
                 {
-                    int h1 = HexToInt(value[pos + 1]);
-                    int h2 = HexToInt(value[pos + 2]);
+                    char ch = value[pos];
 
-                    if (h1 >= 0 && h2 >= 0)
-                    {     // valid 2 hex chars
-                        byte b = (byte)((h1 << 4) | h2);
-                        pos += 2;
+                    if (ch == '+')
+                    {
+                        ch = ' ';
+                    }
+                    else if (ch == '%' && pos < count - 2)
+                    {
+                        int h1 = HexToInt(value[pos + 1]);
+                        int h2 = HexToInt(value[pos + 2]);
 
-                        // don't add as char
-                        helper.AddByte(b);
-                        continue;
+                        if (h1 >= 0 && h2 >= 0)
+                        {     // valid 2 hex chars
+                            byte b = (byte)((h1 << 4) | h2);
+                            pos += 2;
+
+                            // don't add as char
+                            helper.AddByte(b);
+                            continue;
+                        }
                     }
+
+                    if ((ch & 0xFF80) == 0)
+                        helper.AddByte((byte)ch); // 7 bit have to go as bytes because of Unicode
+                    else
+                        helper.AddChar(ch);
                 }
 
-                if ((ch & 0xFF80) == 0)
-                    helper.AddByte((byte)ch); // 7 bit have to go as bytes because of Unicode
-                else
-                    helper.AddChar(ch);
+                return helper.GetString();
+            }
+            finally
+            {
+                // We don't use `using` since UrlDecoder is a struct
+                // and we want to avoid a boxing conversion
```

Comment:
Whoops, forgot that I had avoided making `UrlDecoder` implement `IDisposable`. Will fix.